apps/heureka/src/routes/services/index.tsx

```diff
@@ -3,28 +3,81 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import React from "react"
-import { createFileRoute, getRouteApi } from "@tanstack/react-router"
+import { createFileRoute } from "@tanstack/react-router"
 import { z } from "zod"
 import { Services } from "../../components/Services"
 import { sanitizeSearchParam } from "../../utils"
+import { SELECTED_FILTER_PREFIX } from "../../constants"
+import { fetchServices } from "../../api/fetchServices"
+import { fetchServicesFilters } from "../../api/fetchServicesFilters"
+import { extractFilterSettingsFromSearchParams, getInitialFilters } from "../../components/Services/utils"
 
-const servicesSearchSchema = z.object({
-  service: z
-    .string()
-    .transform((val) => sanitizeSearchParam(val))
-    .optional(),
-})
+// Schema for validating and transforming search parameters related to /services page.
+const servicesSearchSchema = z
+  .object({
+    service: z
+      .string()
+      .transform((val) => sanitizeSearchParam(val))
+      .optional(),
+    searchTerm: z
+      .string()
+      .transform((val) => sanitizeSearchParam(val))
+      .optional(),
+  })
+  .catchall(
+    z.preprocess(
+      (val, ctx) => {
+        // Only allow keys that start with SELECTED_FILTER_PREFIX
+        if (ctx.path.length > 0 && typeof ctx.path[0] === "string" && !ctx.path[0].startsWith(SELECTED_FILTER_PREFIX)) {
+          return undefined
+        }
+        return val
+      },
+      z.union([z.string(), z.array(z.string())])
+    )
+  )
+
+export type ServicesSearchParams = z.infer<typeof servicesSearchSchema>
 
 export const Route = createFileRoute("/services/")({
   validateSearch: servicesSearchSchema,
-  component: RouteComponent,
-})
+  loaderDeps: ({ search }) => {
+    const { service, ...rest } = search
+    return rest
+  },
+  shouldReload: false, // Only reload the route when the user navigates to it or when deps change
+  beforeLoad: ({ context: { appProps }, search }) => {
+    const filterSettings = extractFilterSettingsFromSearchParams(search)
+    return {
+      filterSettings:
+        // Filters from the URL always have preference over initial filters
```

Comment:
(Question)
That logic could be confusing, the nice implementation but would like more clarification on it